AK/Format.h

```diff
@@ -35,6 +37,155 @@
 #    include <stdio.h>
 #endif
 
+#ifndef DBGLN_NO_COMPILETIME_FORMAT_CHECK
+// Note: Clang 12 adds support for CTAD, this would fail with any version prior to that.
+#    if defined(__clang__) && __clang_major__ < 12
+#        define DBGLN_NO_COMPILETIME_FORMAT_CHECK
+#    endif
+#endif
+
+#ifndef DBGLN_NO_COMPILETIME_FORMAT_CHECK
+namespace {
+
+template<size_t N>
+consteval auto extract_used_argument_index(const char (&fmt)[N], size_t specifier_start_index, size_t specifier_end_index, size_t& next_implicit_argument_index)
+{
+    struct {
+        size_t index_value { 0 };
+        bool saw_explicit_index { false };
+    } state;
+    for (size_t i = specifier_start_index; i < specifier_end_index; ++i) {
+        auto c = fmt[i];
+        if (c > '9' || c < '0')
+            break;
+
+        state.index_value *= 10;
+        state.index_value += c - '0';
+        state.saw_explicit_index = true;
+    }
+
+    if (!state.saw_explicit_index)
+        return next_implicit_argument_index++;
+
+    return state.index_value;
+}
+
+// FIXME: We should rather parse these format strings at compile-time if possible.
```

Comment:
I was dreaming of moving the parsing out of `vformat` into `VariadicFormatParams`, that would make this "free". Maybe I'll get around and do this at some point.

Reply:
Perhaps it could be a standalone constexpr function?
For that matter, perhaps we could make `GenericLexer` constexpr, or write another constexpr one that could handle being used to parse format strings?

With this PR we have a constexpr-enabled string (even though it's extremely primitive), so building on that shouldn't be so hard; I don't think `GenericLexer` does anything that can't be done in compile-time.

I'll cc you if I get to doing this before you do ^^

Reply:
I spent today trying to do this (while keeping the interface similar-ish), but it doesn't seem very doable.
I'm not sure how (or even _whether_) the folks over at `{fmt}` did this, since the API we're trying to emulate is not constexpr to begin with...

we could probably this a lot more simple by only having one "formatter" - i.e. the standard formatter, and completely separating the parse/format steps, which would then leave us with a metric tonne of string handling at compiletime. which is a painful experience...

Reply:
If this is "only" about verification and consistency checking, we could make this a separate lint script.

Reply:
@BenWiederhake not really, parsing at compiletime gives us two main things, one being that we don't have to parse at runtime (so faster formatting), and the other being format string validation and such.
we _could_ have two implementations of the parser, one being constexpr, which would entirely sidestep this issue - but that's just a pain.

I personally prefer to have the code not compile at all than to have a linter scream at me.